services/hierarchy/src/main/kotlin/ProductService.kt

```diff
@@ -100,4 +102,60 @@ class ProductService(
     ): Product = db.dbQuery {
         productRepository.update(productId, name, description)
     }
+
+    /**
+     * Add a user to one of the three groups that grant roles and permissions on product level.
+     */
+    suspend fun addUserToGroup(
+        username: String,
+        productId: Long,
+        groupId: String
+    ) {
+        getProduct(productId)
+            ?: throw ResourceNotFoundException("Product with productId '$productId' not found.")
+
+        val productRole = try {
+            ProductRole.valueOf(groupId.uppercase().removeSuffix("S"))
+        } catch (e: IllegalArgumentException) {
+            throw ResourceNotFoundException(
+                "Group with groupId '$groupId' not found. Must be one of ${ProductRole.entries.toJoinedString()}",
+                e
+            )
+        }
+
+        val groupName = productRole.groupName(productId)
+
+        // As the AuthorizationService does not distinguish between technical exceptions (e.g. cannot connect to
+        // Keycloak) and business exceptions (e.g. user not found), we can't do special exception handling here
+        // and just let the exception propagate.
+        authorizationService.addUserToGroup(username, groupName)
+    }
+
+    /**
+     * Remove a user from a group that grant roles and permissions on product level.
+     */
+    suspend fun removeUserFromGroup(
+        username: String,
+        productId: Long,
+        groupId: String
+    ) {
+        getProduct(productId)
+            ?: throw ResourceNotFoundException("Product with productId '$productId' not found.")
+
+        val productRole = try {
+            ProductRole.valueOf(groupId.uppercase().removeSuffix("S"))
+        } catch (e: IllegalArgumentException) {
+            throw ResourceNotFoundException(
+                "Group with groupId '$groupId' not found. Must be one of ${ProductRole.entries.toJoinedString()}",
+                e
+            )
+        }
+
+        val groupName = productRole.groupName(productId)
+
+        // As the AuthorizationService does not distinguish between technical exceptions (e.g. cannot connect to
+        // Keycloak) and business exceptions (e.g. user not found), we can't do special exception handling here
+        // and just let the exception propagate.
+        authorizationService.removeUserFromGroup(username, groupName)
+    }
```

Comment:
There should also be tests for the new functions, similar to those in `OrganizationServiceTest`.